### PR TITLE
[DDW-634] Improve switching application menu screens

### PR DESCRIPTION
**Pull request by DeeJayElly**
This PR CHANGES switching between application menu screens by rechecking if any of the previous screens is already open or not, if yes, then closing it first, before opening new screen.

**NOTE:**
In case the user tries to open the same dialog twice nothing should happen - the existing dialog, in that case, remains open...

## Todos

- [x] Add CHANGELOG entry

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1560167761000400)

- [ ] Test functionality of opening/closing About Us, Daedalus Diagnostics, Block Consolidation & Ada Redemption screens. Opening of new screen should close the previous open screen.
- [ ] Test Ada Redemption screen opening from different locations.
---

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes and shows some screenshots or animated GIFs of important UI changes
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

File: source/renderer/app/App.js [hunk 1]
```diff
@@ -45,10 +46,16 @@ export default class App extends Component<{
...
                 {mobxDevTools}
-                {isDaedalusDiagnosticsDialogOpen && (
+                {activeDialog === 'daedalusDiagnostics' && (
                   <DaedalusDiagnosticsDialog />
                 )}
-                {isAboutDialogOpen && <AboutDialog />}
+                {activeDialog === 'about' && <AboutDialog />}
+                {activeDialog === 'blockConsolidation' && (
+                  <BlockConsolidationStatusDialog
+                    stores={stores}
```
Comment (nikolaglumac) [inline source/renderer/app/App.js]: @DeeJayElly are you sure we need to provide stores and actions to the container? We don't do that for other dialogs...

File: source/renderer/app/stores/AppStore.js [hunk 2]
```diff
@@ -13,39 +12,38 @@ import {
...
 
 export default class AppStore extends Store {
   @observable error: ?LocalizableError = null;
-  @observable isAboutDialogOpen = false;
-  @observable isDaedalusDiagnosticsDialogOpen = false;
   @observable isDownloadNotificationVisible = false;
   @observable gpuStatus: ?GpuStatus = null;
   @observable numberOfEpochsConsolidated: number = 0;
   @observable previousRoute: string = ROUTES.ROOT;
+  @observable activeDialog: string = '';
```
Comment (nikolaglumac) [inline source/renderer/app/stores/AppStore.js]: @DeeJayElly can we please have a special `Dialog` (or some better name) type and use null for default when there is no active dialog?

**Comment by nikolaglumac** [comment]
@DeeJayElly I have set this PR back to WIP state until the changes I have requested are implemented.

File: source/renderer/app/stores/AppStore.js (hunk 2 above)
Comment (DeeJayElly) [reply source/renderer/app/stores/AppStore.js]: @nikolaglumac done

**Comment by DeeJayElly** [comment]
> @DeeJayElly please check the 2 comments I have left and remove the unused routes.

Done

File: source/renderer/app/App.js (hunk 1 above)
Comment (DeeJayElly) [reply source/renderer/app/App.js]: @nikolaglumac done

**Comment by nikolaglumac** [comment]
@DeeJayElly please remove the unused route from here too: https://github.com/input-output-hk/daedalus/blob/chore/ddw-634-improve-switching-application-menu-screens/source/renderer/app/routes-config.js#L5

**Comment by DeeJayElly** [comment]
> @DeeJayElly please remove the unused route from here too: https://github.com/input-output-hk/daedalus/blob/chore/ddw-634-improve-switching-application-menu-screens/source/renderer/app/routes-config.js#L5

Done

File: source/renderer/app/stores/AppStore.js (hunk 2 above)
Comment (nikolaglumac) [reply source/renderer/app/stores/AppStore.js]: @DeeJayElly I am afraid you did not understand me here - I did not want to add `current` and type that - I wanted the `ActiveScreenType` to not be a string but rather a list of possible dialogs such as that this line reads:
```
@observable activeDialog: ActiveDialogName = null;
```
...and then `ActiveDialogName` is defined like:
```
export type ActiveDialogName =
  | 'about'
  | 'adaRedemption'
  | 'blockConsolidation'
  | 'daedalusDiagnostics'
  | null;
```

**Comment by nikolaglumac** [comment]
@DeeJayElly looks like there are some flow errors too :(

**Comment by DeeJayElly** [comment]
> @DeeJayElly looks like there are some flow errors too :(

@nikolaglumac fixed, all good now

File: source/renderer/app/stores/AppStore.js (hunk 2 above)
Comment (DeeJayElly) [reply source/renderer/app/stores/AppStore.js]: yep i changed that part now. Its more cleaner and packed in external type & const.

**Pull request merged**